vm/systemSmartContracts/eei.go

```diff
@@ -0,0 +1,201 @@
+		return actualBalance
+	}
+
+	balance, err := host.blockChainHook.GetBalance(addr)
+	if err != nil {
+		return nil
+	}
+
+	host.outputAccounts[strAdr] = &vmcommon.OutputAccount{
+		Balance:      big.NewInt(0).Set(balance),
+		BalanceDelta: big.NewInt(0),
+		Address:      addr}
+
+	return balance
+}
+
+// Transfer handles any necessary value transfer required and takes
+// the necessary steps to create accounts
+func (host *vmContext) Transfer(
+	destination []byte,
+	sender []byte,
+	value *big.Int,
+	input []byte,
+) error {
+
+	senderAcc, ok := host.outputAccounts[string(sender)]
+	if !ok {
+		senderAcc = &vmcommon.OutputAccount{
+			Address:      sender,
+			BalanceDelta: big.NewInt(0),
+			Balance:      big.NewInt(0),
+		}
+		host.outputAccounts[string(senderAcc.Address)] = senderAcc
+	}
+
+	destAcc, ok := host.outputAccounts[string(destination)]
+	if !ok {
+		destAcc = &vmcommon.OutputAccount{
+			Address:      destination,
+			BalanceDelta: big.NewInt(0),
+			Balance:      big.NewInt(0),
+		}
+		host.outputAccounts[string(destAcc.Address)] = destAcc
+	}
+
+	senderAcc.BalanceDelta = big.NewInt(0).Sub(senderAcc.BalanceDelta, value)
+	destAcc.BalanceDelta = big.NewInt(0).Add(destAcc.BalanceDelta, value)
+
+	return nil
+}
+
+// CleanCache cleans the current vmContext
+func (host *vmContext) CleanCache() {
+	host.storageUpdate = make(map[string]map[string][]byte, 0)
+	host.selfDestruct = make(map[string][]byte)
+	host.outputAccounts = make(map[string]*vmcommon.OutputAccount, 0)
+	host.output = make([]byte, 0)
+}
+
+// CreateVMOutput adapts vm output and all saved data from sc run into VM Output
+func (host *vmContext) CreateVMOutput() *vmcommon.VMOutput {
+	vmOutput := &vmcommon.VMOutput{}
+	// save storage updates
+	outAccs := make(map[string]*vmcommon.OutputAccount, 0)
+	for addr, updates := range host.storageUpdate {
+		if _, ok := outAccs[addr]; !ok {
+			outAccs[addr] = &vmcommon.OutputAccount{Address: []byte(addr)}
+		}
+
+		for key, value := range updates {
+			storageUpdate := &vmcommon.StorageUpdate{
+				Offset: []byte(key),
+				Data:   value,
+			}
+
+			outAccs[addr].StorageUpdates = append(outAccs[addr].StorageUpdates, storageUpdate)
+		}
+	}
+
+	// add balances
+	for addr, outAcc := range host.outputAccounts {
+		if _, ok := outAccs[addr]; !ok {
+			outAccs[addr] = &vmcommon.OutputAccount{}
+		}
+
+		outAccs[addr].Address = outAcc.Address
+		outAccs[addr].BalanceDelta = outAcc.BalanceDelta
+
+		if len(outAcc.Code) > 0 {
+			outAccs[addr].Code = outAcc.Code
+		}
+		if outAcc.Nonce > 0 {
+			outAccs[addr].Nonce = outAcc.Nonce
+		}
+	}
+
+	// add self destructed contracts
+	for addr := range host.selfDestruct {
+		vmOutput.DeletedAccounts = append(vmOutput.DeletedAccounts, []byte(addr))
+	}
+
+	// save to the output finally
+	for _, outAcc := range outAccs {
+		vmOutput.OutputAccounts = append(vmOutput.OutputAccounts, outAcc)
+	}
+
+	vmOutput.GasRemaining = big.NewInt(0)
```

Comment:
Is this supposed to be this way? We might have a fixed gas for these VM operations. It is not fair to refund 0 when calling a protocol SC func as you might want to put a large gas limit so you are assured that the operation is done.

Reply:
currently there are no gas usage on systemVM, when it is added, this line will be changed.